### PR TITLE
sql-parser: add consistency topic format to Kafka connector

### PR DESCRIPTION
**Pull request by JLDLaughlin**
A redo of #7592! 

This change updates the `CREATE SINK` syntax to accept new `CONSISTENCY TOPIC` and `CONSISTENCY FORMAT` parameters. This change is needed to expand our sink formats.

The new Kafka connector syntax looks like:
Screen Shot 2021-07-28 at 9 26 22 PM

It is definitely repetitive! I repeat `CONSISTENCY` for `FORMAT` to disambiguate from the general topic's `FORMAT`, which is parsed next.

### CODE REVIEW

**Comment by JLDLaughlin** [comment]
Also, I will add a release note about all of this (JSON sinks for Kafka, new syntax details) when the last PR lands!

File: src/sql-parser/src/ast/defs/ddl.rs [hunk 1]
```diff
@@ -371,6 +371,8 @@ pub enum Connector {
         broker: String,
         topic: String,
         key: Option<Vec<Ident>>,
+        consistency_topic: Option<String>,
+        consistency_topic_format: Option<Format<T>>,
```
Comment (benesch) [inline src/sql-parser/src/ast/defs/ddl.rs:375]: Tiny nit: it's probably worth having a `KafkaConsistency` struct that you can optionalize separately! That way you won't be able to represent a consistency topic format with no topic name:

```rust
enum Connector {
    Kafka {
        // ...
        consistency: Option<KafkaConsistency>
    }
}

struct KafkaConsistency {
    topic: String,
    topic_format: Option<Format<T>>
}
```

File: doc/user/sql-grammar/sql-grammar.bnf [hunk 2]
```diff
@@ -321,6 +321,7 @@ sink_kafka_connector ::=
     'KAFKA BROKER' host 'TOPIC' topic-prefix
     ('KEY' '(' key_column ( ',' key_column )* ')')?
     ('WITH' '(' ( field '=' val ) ( ( ',' field '=' val ) )* ')')?
+    ('CONSISTENCY' 'TOPIC' topic ('CONSISTENCY' 'FORMAT' 'AVRO USING' 'CONFLUENT SCHEMA REGISTRY' url)?)?
```
Comment (benesch) [inline doc/user/sql-grammar/sql-grammar.bnf:324]: I think this wants to be one line higher?

File: src/sql/src/plan/statement/ddl.rs [hunk 3]
```diff
@@ -1539,8 +1540,14 @@ pub fn plan_create_sink(
...
+            broker,
+            topic,
+            consistency_topic_format,
+            ..
+        } => kafka_sink_builder(
             format,
+            consistency_topic_format,
             &mut with_options,
             broker,
             topic,
```
Comment (benesch) [inline src/sql/src/plan/statement/ddl.rs:1553]: Are you planning on bringing that same check for if the consistency topic is specified twice into here?

File: src/sql-parser/src/ast/defs/ddl.rs (hunk 1 above)
Comment (benesch) [reply src/sql-parser/src/ast/defs/ddl.rs:375]: (Er, that is, if it is true that you can't specify the format without specifying a topic.)

File: src/sql/src/plan/statement/ddl.rs (hunk 3 above)
Comment (JLDLaughlin) [reply src/sql/src/plan/statement/ddl.rs:1553]: Yes! I am

File: src/sql-parser/src/ast/defs/ddl.rs (hunk 1 above)
Comment (JLDLaughlin) [reply src/sql-parser/src/ast/defs/ddl.rs:375]: That's true! Fixing

File: doc/user/sql-grammar/sql-grammar.bnf (hunk 2 above)
Comment (JLDLaughlin) [reply doc/user/sql-grammar/sql-grammar.bnf:323]: Yes! Thank you -- diagram in the PR description has been updated as well.

File: src/sql/src/plan/statement/ddl.rs [hunk 4]
```diff
@@ -1259,6 +1260,9 @@ fn kafka_sink_builder(
         Some(Value::String(topic)) => Some(topic),
         Some(_) => bail!("consistency_topic must be a string"),
     };
+    if consistency_topic.is_some() && consistency.is_some() {
+        bail!("cannot specify consistency_topic and CONSISTENCY TOPIC simultaneously");
+    }
```
Comment (benesch) [inline src/sql/src/plan/statement/ddl.rs:1266]: We should do something with `consistency` here right? Are you planning to add that to this PR? If not, probably worth dropping an `unsupported!` here?

File: doc/user/content/sql/create-sink.md [hunk 5]
```diff
@@ -42,9 +42,11 @@ Field | Use
 ------|-----
 **KAFKA BROKER** _host_ | The Kafka broker's host name without the security protocol, which is specified by the [`WITH` options](#with-options).) If you wish to specify multiple brokers (bootstrap servers) as an additional safeguard, use a comma-separated list. For example: `localhost:9092, localhost:9093`.
 **TOPIC** _topic&lowbar;prefix_ | The prefix used to generate the Kafka topic name to create and write to.
+**KEY (** _key&lowbar;column&lowbar;list_ **)** | An optional list of columns to use for the Kafka key. If unspecified, the Kafka key is left unset. {{< version-added v0.5.1 />}}
 **WITH OPTIONS (** _option&lowbar;_ **)** | Options affecting sink creation. For more details see [`WITH` options](#with-options).
+**CONSISTENCY TOPIC** _consistency&lowbar;topic_ | The Kafka topic name to write [consistency metadata](#consistency-metadata) to.
+**CONSISTENCY FORMAT** _format_ | The format of the Kafka consistency topic, defaults to Avro for Avro-encoded sinks.
```
Comment (benesch) [inline doc/user/content/sql/create-sink.md:48]: Can you add one of those `{{< version-added >}}` tags to the docs here (see `KEY` for an example)?

**Review by benesch — APPROVED** [review]
LGTM but I am confused about draft-ness! Are you planning to push more changes here? If not see comment within about restricting the use of this syntax.

**Comment by JLDLaughlin** [comment]
I forgot it was a draft! I will fix those last two things and merge in the morning. (I have a separate PR planned for actually using the new consistency struct.) Thanks for the quick reviews, Nikhil!

File: doc/user/content/sql/create-sink.md [hunk 6]
```diff
@@ -42,9 +42,11 @@ Field | Use
 ------|-----
 **KAFKA BROKER** _host_ | The Kafka broker's host name without the security protocol, which is specified by the [`WITH` options](#with-options).) If you wish to specify multiple brokers (bootstrap servers) as an additional safeguard, use a comma-separated list. For example: `localhost:9092, localhost:9093`.
 **TOPIC** _topic&lowbar;prefix_ | The prefix used to generate the Kafka topic name to create and write to.
+**KEY (** _key&lowbar;column&lowbar;list_ **)** | An optional list of columns to use for the Kafka key. If unspecified, the Kafka key is left unset. {{< version-added v0.5.1 />}}
 **WITH OPTIONS (** _option&lowbar;_ **)** | Options affecting sink creation. For more details see [`WITH` options](#with-options).
+**CONSISTENCY TOPIC** _consistency&lowbar;topic_ | The Kafka topic name to write [consistency metadata](#consistency-metadata) to.
```
Comment (aljoscha) [inline doc/user/content/sql/create-sink.md:47]: I think we should also forward the description from `consistency_topic`:

```
Makes the sink emit additional consistency metadata. Only valid for Kafka sinks. If reuse_topic is true, a default consistency_topic will be used when not explicitly set. The default consistency topic name is formed by appending -consistency to the output topic name.
```

And further below there is a section that mentions "Instead, the topic name specified via the consistency_topic option is used as is.". We should update that to be agnostic of how the consistency topic is specified.

Comment (JLDLaughlin) [reply doc/user/content/sql/create-sink.md:47]: > We should update that to be agnostic of how the consistency topic is specified.

Totally! Right now I'm not doing anything with the `CONSISTENCY TOPIC` information. I'll put that docs change in the next PR when I actually use it, if that's okay with you!

Comment (JLDLaughlin) [reply doc/user/content/sql/create-sink.md:47]: But I made the first change!

Comment (aljoscha) [reply doc/user/content/sql/create-sink.md]: Yes, that sounds good!

**Pull request merged**